docs/how-to/write-plans.md

```diff
@@ -1,173 +1,81 @@
+## Injecting Devices
 
-    Allowed argument types for Pydantic BaseModels include the primitives, types that extend `BaseModel` and `dict`\ s, `list`\ s  and other `sequence`\ s of supported types. Blueapi will deserialise these types from JSON, so `dict`\ s should use `str` keys.
+Some plans are created for specific sets of devices, or will almost always be used with the same devices, it is useful to be able to specify defaults. [Dodal makes this easy with its factory functions](https://diamondlightsource.github.io/dodal/main/how-to/include-devices-in-plans.html).
 
-## Injecting defaults
-
-Often when writing a plan, it is known which device the plan will mostly or always be run with, but at the time of writing the plan the device object has not been instantiated: dodal defines device factory functions, but these cannot be injected as default arguments to plans.
-
-Dodal defines an `inject` function which bypasses the type checking of the constructed schemas, defering to the blueapi contexting fetching of the device when the plan is imported. This allows defaulting devices, so long as there is a device of that name in the context which conforms to the type annotation.
-
-``` 
-   def touch_synchrotron(sync: Synchrotron = inject("synchrotron")) -> MsgGenerator:
-       # There is only one Synchrotron device, so we know which one it will always be.
-       # If there is no device named "synchrotron" in the blueapi context, it will except.
-       sync.specific_function()
-       yield from {}
-```
-
-### Metadata
+## Injecting Metadata
 
 The bluesky event model allows for rich structured metadata to be attached to a scan. To enable this to be used consistently, blueapi encourages a standard form.
 
-> **_NOTE:_** Plans **should** include `metadata` as their final argument, if they do it **must** have the type Optional[Mapping[str, Any]], `and a default of None <https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python>`__\, with the plan defaulting to an empty dict if passed `None`. If the plan calls to a stub/plan which takes metadata, the plan **must** pass down its metadata, which may be a differently named argument.
+Plans ([as opposed to stubs](../explanations/plans.md)) **should** include `metadata` as their final parameter, if they do it **must** have the type `Mapping[str, Any] | None`, [and a default of None](https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python). If the plan calls to a stub/plan which takes metadata, the plan **must** pass down its metadata, which may be a differently named parameter.
 
-```
-   def pass_metadata(x: Movable, metadata: Optional[Mapping[str, Any]] = None) -> MsgGenerator:
-       yield from bp.count{[x], md=metadata or {}}
+```python
+def pass_metadata(x: Movable, metadata: Mapping[str, Any] | None = None) -> MsgGenerator:
```

Comment:
I think we need to tighten up Mapping -> dict to behave well with bp and bps usage.

Reply:
Sorry, not sure I follow, what issues are you seeing?

Reply:
bluesky.plans have their equivalent md argument (which we should use when possible) is `CustomPlanMetadata = dict[str, Any]`: we therefore cannot pass a `Mapping[str, Any]` down to it: `Mapping` does not allow assignment, where `dict` does. We also sometimes need to mutate metadata in a "sub"plan before passing it to a "super"plan.

Reply:
This very plan would fail type checking for the issue. We can quite neatly then have the required parts of the plan be from bluesky.utils:

```python
from bluesky.utils import CustomPlanMetadata, MsgGenerator

def foo(metadata: CustomPlanMetadata | None = None) -> MsgGenerator:
    yield from {}
```

Reply:
I'm happy to change it, please feel free to read or ignore the small rant below.

The reason it's `dict` is because it's mutated upstream:

```python
def my_plan(params: ..., md: dict[str, Any]):
    ...
    md["plan_name"] = my_plan.__name__
    yield from ...
```

It could instead be kept immutable:

```python
def my_plan(params: ..., md: Mapping[str, Any]):
    ...
    md = {**md, **{"plan_name": my_plan.__name__}}
    yield from ...
```

Immutable = good, so I would prefer we change upstream, but the verbosity of that solution may be unpalletable.